### PR TITLE
Replace tarpc

### PR DESCRIPTION
**Pull request by freshstrangemusic**
This replaces the tarpc usage with our own bespoke RPC that I had written some months ago.

At a high level, the messages are sent as length-delimited JSON blobs serialized with Serde. `fxrecorder` sends `RecorderMessages` to `fxrunner`, which replies with `RunnerMessages`. There are some *unfortunate* type signatures in `libfxrecord::net::proto`, due to lack of support for higher-kinded types in Rust. Additionally, there is a *lot* of codegen in `libfxrecord::net::message` via macros that I've attempted to document.

### CODE REVIEW

File: libfxrecord/src/lib.rs [hunk 1]
```diff
@@ -45,7 +45,7 @@ where
         .and_then({
             let log = log.clone();
             move |config| {
-                let rt = Runtime::new().expect("could not get tokio runtime");
+                let mut rt = Runtime::new().expect("could not get tokio runtime");
```
Comment (squarewave) [inline libfxrecord/src/lib.rs]: Just curious - did it actually compile before the `mut`? It doesn't look like `block_on` ever didn't require a `mut` reference.

File: libfxrecord/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::error::Error;
+use std::fmt::{Debug, Display};
+
+use derive_more::Display;
+
+/// An error that consists only of a message and no state.serde
```
Comment (squarewave) [inline libfxrecord/src/error.rs]: "state.serde"? Leftovers?

File: libfxrecord/src/net/message.rs [hunk 3]
```diff
@@ -0,0 +1,348 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Message types used throughout `fxrunner` and `fxrecorder`.
+//!
+//! This module consists of various helper traits and and structures, as well
```
Comment (squarewave) [inline libfxrecord/src/net/message.rs]: "and and"

File: docs/protocol.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+The recorder will send a `Handshake` message to the runner, indicating that
+it should restart. The runner replies with a `HandshakeReply` with the status
+of the restart operation. They then disconnect and the recorder waits for the
+runner to restart.
+
+> ![](/docs/diagrams/handshake.png)
+>
+> Figure 1: Handshake
+
+If something goes wrong with the handhshake on the runner's end (such as a
```
Comment (squarewave) [inline docs/protocol.md]: "handhshake"

**Review by squarewave — DISMISSED** [review]
LGTM - just a few typo corrections in the comments!

File: libfxrecord/src/lib.rs (hunk 1 above)
Comment (freshstrangemusic) [reply libfxrecord/src/lib.rs]: I originally wrote this against tokio 0.2.0-alpha-6, which did not require a `mut` and missed addressing this in the [`Update Dependencies`](https://github.com/mozilla/fxrecord/commit/b9afe778495e218089e61cbf6849a9ce37b8d93f). It was rolled into the [tarpc commit](https://github.com/mozilla/fxrecord/commit/11254300fd5cdeb5de28996d73c62596f23ca14b#diff-4466f9bbc1f9fa9b9b98189e61664e1fR49) but that was reverted

File: libfxrecord/src/error.rs (hunk 2 above)
Comment (freshstrangemusic) [reply libfxrecord/src/error.rs]: Sometimes VSCode decides that it should autocomplete  based on a word in a comment and I end up with stuff like this.

**Review by keeferrourke — APPROVED** [review]

**Review by keeferrourke — APPROVED** [review]
Still LGTM :)

**Review by squarewave — APPROVED** [review]
👍

**Pull request merged**